langchain4j-oracle/src/main/java/dev/langchain4j/store/embedding/oracle/OracleEmbeddingStore.java

```diff
@@ -1,542 +1,592 @@
-                    }
-
-                    matches.add(new EmbeddingMatch<>(score, id, embedding, textSegment));
+        ensureNotNull(request, "request");
+
+        SQLFilter sqlFilter = SQLFilters.create(request.filter(), METADATA_KEY_MAPPER);
+        final int maxResults = request.maxResults();
+
+        // In a 23.4 build of Oracle Database, ORA-06553 will result if the distance column is referenced in the WHERE
+        // clause. For this reason, the minScore filtering will happen locally. There are workarounds, such as binding
+        // the VECTOR twice, or using a sub-select. These can be implemented if they are proven to offer a significant
+        // performance improvement.
+        try (Connection connection = dataSource.getConnection();
+             PreparedStatement query = connection.prepareStatement(
+                     "SELECT embedding <=> ? distance, id, embedding, content, metadata"
+                         + " FROM " + tableName
+                         + sqlFilter.asWhereClause()
+                         + " ORDER BY distance"
+                         + " FETCH FIRST " + maxResults + " ROWS ONLY")
+        ) {
+            query.setObject(1, request.queryEmbedding().vector(), OracleTypes.VECTOR_FLOAT32);
+            sqlFilter.setParameters(query, 2);
+            query.setFetchSize(maxResults);
+
+            // Calls to defineColumnType reduce the number of network requests. When Oracle JDBC knows that it is
+            // fetching VECTOR, CLOB, and/or JSON columns, the first request it sends to the database can include a LOB
+            // prefetch size (VECTOR and JSON are value-based-lobs). If defineColumnType is not called, then JDBC needs
+            // to send an additional request with the LOB prefetch size, after the first request has the database
+            // respond with the column data types.
+            OracleStatement oracleStatement = query.unwrap(OracleStatement.class);
+            oracleStatement.defineColumnType(1, OracleTypes.BINARY_DOUBLE);
+            oracleStatement.defineColumnType(2, OracleTypes.VARCHAR);
+            oracleStatement.defineColumnType(3, OracleTypes.VECTOR_FLOAT32, 524308);
+            oracleStatement.defineColumnType(4, OracleTypes.CLOB, Integer.MAX_VALUE);
+            oracleStatement.defineColumnType(5, OracleTypes.JSON, Integer.MAX_VALUE);
+
+            List<EmbeddingMatch<TextSegment>> matches = new ArrayList<>(maxResults);
+            try (ResultSet resultSet = query.executeQuery()) {
+                while (resultSet.next()) {
+
+                    // Convert the cosine distance between 0 and 2 into a score between 1 and 0.
+                    double score = 1d - (resultSet.getDouble("distance") / 2d);
+
+                    // Local filtering of the minScore. See note about ORA-06553 above.
+                    if (score < request.minScore())
+                        continue;
```

Comment:
Can't you break here instead of continue ? Since your results are ordered by distances, all results coming after would have a worse score, wouldn't they ?

Reply:
True! Thank you.

Reply:
Fixed in latest commit. Thank you.